### PR TITLE
Fix more `DM_DEFAULT_ENCODING` SpotBugs violations

### PR DESCRIPTION
**Pull request by basil**
Fixes more `DM_DEFAULT_ENCODING` SpotBugs violations by using an explicit character set.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java
Comment (basil) [inline core/src/main/java/hudson/FilePath.java:1600]: Just a few lines above, the Javadoc states (emphasis mine):

> Creates a temporary file in this directory […] and set the contents to the given text (**encoded in the platform default encoding**)

File: core/src/main/java/hudson/lifecycle/WindowsServiceLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/WindowsServiceLifecycle.java:115]: A Windows-specific file, so we cannot assume UTF-8.

File: core/src/main/java/hudson/util/TextFile.java
Comment (basil) [inline core/src/main/java/hudson/util/TextFile.java:115]: `TextFileTest` uses `Charset.defaultCharset` when writing the test string, so it seems only reasonable that we would also use it when reading the same string.

File: core/src/main/java/jenkins/diagnosis/HsErrPidList.java
Comment (basil) [inline core/src/main/java/jenkins/diagnosis/HsErrPidList.java:127]: I couldn't find any documentation on the encoding of `hs_err_pid`, but it seems reasonable to preserve existing behavior here.

**Review by StefanSpieker — APPROVED** [review]

**Comment by basil** [comment]
Passed testing in [Windows CI](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6098/2/), [BOM](https://ci.jenkins.io/job/Tools/job/bom/job/PR-776/3/), and interactive usage (setup wizard & building Pipeline job) on Windows & Linux.

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:157]: Agent log files should be in UTF-8.

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:177]: ditto

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java
Comment (basil) [reply core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:157]: Here and elsewhere, please provide reasoning when you make requests of me.

File: core/src/main/java/hudson/util/TextFile.java
Comment (basil) [reply core/src/main/java/hudson/util/TextFile.java:115]: But it does match https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/util/TextFile.java#L176-L181 so by the same reasoning I suggest leaving the code as-is. Fair enough?

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:157]: Any plugin can write text to these logs, and it is intended solely for use by Jenkins, so there is no reason to allow non-UTF-8 encoding.

Looks like some other places need to be fixed; at least: https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/slaves/SlaveComputer.java#L155 https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/slaves/SlaveComputer.java#L623 https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/model/Computer.java#L335 https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/model/Computer.java#L343

Comment (basil) [reply core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:157]: A bigger change than what I am trying to achieve in this PR. For now I have added a comment matching https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/model/Computer.java#L335 to document the status quo.

File: core/src/main/java/hudson/util/TextFile.java
Comment (jglick) [reply core/src/main/java/hudson/util/TextFile.java:115]: These are secondary methods. `read` uses UTF-8 so everything else should be updated accordingly.

Comment (basil) [reply core/src/main/java/hudson/util/TextFile.java:115]: commit 14d980c437 FTR

Comment (basil) [reply core/src/main/java/hudson/util/TextFile.java]: > `read` uses UTF-8 so everything else should be updated accordingly.

Pre-existing issue, but never mind that. Done.

Comment (basil) [reply core/src/main/java/hudson/util/TextFile.java]: Actually I am not comfortable taking on this risk changing the behavior of an existing public API. Reverted.

Comment (jglick) [reply core/src/main/java/hudson/util/TextFile.java:115]: Then at least leave comments noting that the behavior of methods in this class is inconsistent. Not even sure if these non-`read` methods are used.

Comment (basil) [reply core/src/main/java/hudson/util/TextFile.java:127]: Opted for a more conservative approach of deprecating the methods that use the default charset in favor of non-deprecated versions that take an explicit charset.

File: core/src/main/java/jenkins/diagnosis/HsErrPidList.java
Comment (jtnord) [reply core/src/main/java/jenkins/diagnosis/HsErrPidList.java:127]: source code [looks like platform default](https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/test/hotspot/jtreg/runtime/ErrorHandling/TestCrashOnOutOfMemoryError.java#L108-L121) or the test is wrong :)

File: core/src/main/java/hudson/util/TextFile.java
Comment (jtnord) [reply core/src/main/java/hudson/util/TextFile.java:128]: > read uses UTF-8 so everything else should be updated accordingly.

as does `write` and `lines` so these `head`/`tail` are bogus as is the comment on `fastTail` as the only possibility is for this to every be a `UTF-8` encoded text file?

Thus forcing users to specify a charset for `tail`/`head` to be "correct" when the only correct possibility is "UTF-8" for the encoding of the text in the file seems wrong to me, and the mere existance of methods that take a charset would be a flag (and I note that some already exist!).

Besides that - this code is wrong as it is ignoring the passed in `Charset` and using `Charset.defaultCharset()`

there only seems to be a couple of usages of [`fastTail`](https://github.com/search?q=org%3Ajenkinsci+fastTail&type=code) - 
 `junit` plugin seems to be using this for files that are not written by textFile but existing files.  That seems like an abuse to to me, and changing behaviour here without first fixing that would likely be a regression.

`usage-in-plugins` reports just a few callers of these methods - so it would seem to be better to fix those (if they are similar to junit), and then not take a charset at all and make TextFile always be `UTF-8`?

>flaky-test-handler:1.2.0
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;

>junit:1.53
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;

>mentor-questa-vrm:1.13
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;

**Review by jtnord — COMMENTED** [review]
just concerns around the new API in `TextFile` that makes the class more verbose for very few users.

otherwise LGTM

**Comment by basil** [comment]
As indicated previously, I am _not_ interested in going on another ecosystem-wide hunt and I will _not_ increase the scope of this PR beyond the current scope of the existing public API. I have reverted any and all changes to `TextFile`.

**Comment by jglick** [comment]
That is fine, better to have some warnings fixed than none. Note that as of https://github.com/openjdk/jdk/pull/4733 the whole `DM_DEFAULT_ENCODING` warning will arguably become obsolete anyway.

File: core/src/main/java/jenkins/diagnosis/HsErrPidList.java
Comment (basil) [reply core/src/main/java/jenkins/diagnosis/HsErrPidList.java:127]: Nice find!

**Comment by basil** [comment]
> That is fine, better to have some warnings fixed than none.

Was it your intention to leave this comment without approving this PR? If so, I am having a hard time understanding what the next steps are for landing this.

**Review by jglick — APPROVED** [review]
Sorry, just needed to set aside a few minutes to look over it again.

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
